server/models/questions/forkQuestion.js

```diff
@@ -0,0 +1,90 @@
+const { pool } = require('../database')
+
+/**
+ *
+ * @param {*} param0
+ * @param {String} param0.username
+ * @param {Number} param0.questionId
+ * @return {Promise}
+ */
+
+function forkQuestion({ username, questionId }) {
+  return new Promise((resolve, reject) => {
+    pool.getConnection((err, connection) => {
+      if (err) {
+        return reject(err)
+      }
+      connection.beginTransaction((err) => {
+        if (err) {
+          connection.release()
+          return reject(err)
+        }
+        connection.query(
+          `INSERT INTO questions(creator,type,name,problem_statement,input_format,
+            output_format,constraints,options,correct,difficulty)
+            SELECT creator,type,name,problem_statement,input_format,
+            output_format,constraints,options,correct,difficulty 
+            FROM questions q
+            JOIN questions_editors qe ON q.id=qe.question_id 
+            WHERE q.id=? AND qe.editor=?`,
```

Comment:
Unnecessary join. It would be much slower. Better to use `EXISTS` subquery. The reason being that this EXISTS query would be evaluated only once (reason being it is not a correlated subquery) . In case it is false, it would not look into the `questions` table. And in case it is true, it will simply look only into the `questions` table, instead of performing any JOIN operation.
JOIN would unnecessary merge the 2 large tables (say, both contain 1000-1000 records) and then perform search on that (1000x1000) table, which is not required in our case.

The query should be -
```suggestion
          `INSERT INTO questions(creator,type,name,problem_statement,input_format,
            output_format,constraints,options,correct,difficulty)
            SELECT creator,type,name,problem_statement,input_format,
            output_format,constraints,options,correct,difficulty 
            FROM questions
            WHERE EXISTS(SELECT 1 FROM questions_editors WHERE question_id=? AND editor=?) AND id=?`,
```
And query array would become
```
[questionId, username, questionId]
```

Reply:
I had done the same first..thinking join would execute slower...but then i saw various blogs online like [this one](https://cc.davelozinski.com/sql/fastest-way-to-insert-new-records-where-one-doesnt-already-exist) where they were comparing various methods of insert select. For small number of records, what you said makes complete sense, in fact it join is much slower in that case ( I just found out ), but as we scale up, JOIN is **better**. But for now for small number of records, EXISTS seems fine. I will changes this

Reply:
Actually, the example in that blog is of a correlated subquery, where JOINS are really much better than the correlated subqueries, because correlated subqueries are performed for all the rows of a table

Reply:
Yep. True. Did not notice that.